Spring Cloud Stream Message Based Job Executor Implementation

igdianov opened the pull request:
MessageBasedJobExecutor uses Spring Cloud Stream to emit scheduled jobs as messages into job manager producer channel. Job message payload contains scheduled job id;

The job manager producer channel destination is scoped by runtime bundle's service name in order to restrict handling of job messages within the same group its of instances.

Job message producer uses transaction listener to synchronize sending with Activiti engine transactions.

The job message consumer dynamically creates input channel and binds it to job message producer channel destination to receive job messages. The job consumer must use job handler consumer group for input binding. If there are multiple job consumer instances bound with the same group name, then job messages are load-balanced across those consumer instances so that each message sent by a job message producer is consumed by only a single consumer instance within job message handler group

When job message handler receives job message from input channel, it tries to find existing job entity by job id and delegates job execution to ExecuteAsyncRunnable.

In case exception is raised during job execution, the job handler throws ActivitiException to reject job message. The failed message will be redelivered to retry with configured number of retries.

The following BPMN executions are covered:

![image](https://user-images.githubusercontent.com/20428629/61350085-2e8bad80-a81b-11e9-8463-369b6798f3fb.png)

![image](https://user-images.githubusercontent.com/20428629/61349864-84ac2100-a81a-11e9-8196-1d9d2b21e339.png)

![image](https://user-images.githubusercontent.com/20428629/61349889-95f52d80-a81a-11e9-91a3-0e29b3cebf94.png)

![image](https://user-images.githubusercontent.com/20428629/61349898-9d1c3b80-a81a-11e9-8671-5c80c9f21352.png)

![image](https://user-images.githubusercontent.com/20428629/61349937-bd4bfa80-a81a-11e9-865c-6e0fb31f3423.png)

![image](https://user-images.githubusercontent.com/20428629/61350128-5844d480-a81b-11e9-875a-db2b0787e779.png)

Depends on https://github.com/Activiti/Activiti/pull/2826

Tests Coverage:

- [x] Async Task Job Execution
- [x] Intermediate Timer Catch Event
- [x] Start Timer Event
- [x] Boundary Timer Catch Event
- [x] Externalized Message Job Consumer Properties
- [x] Spring Transaction synchronization
- [x] Multiple Runtime Bundle Replicas
- [x] Spring Application Events support
- [x] Job message headers

igdianov committed:
feat: draft activiti-cloud-services-job-executor module

igdianov committed:
fix: set JobExecutorIT async job count to 100

igdianov committed:
fix: try filter only asyncTask executions in JobExecutorIT test

igdianov committed:
fix: polish JobExecutorIT test

salaboy commented:
@igdianov @erdemedeiros @ryandawsonuk @daisuke-yoshimoto we need to be careful with this. 
I totally agree that we need this but the infrastructure (kubernetes) is already providing a Job/Timers infrastructure. Also https://cloud.google.com/knative/ is the way to go forward for connectors and related out of process jobs/timers execution. While I agree that implementing using Spring Cloud Streams is a good idea for now, we should rely on the infrastructure for such tasks. 

@igdianov I don't see a separated (scalable, out of processs) consumer , which makes me think that we just implemented a local version for now. Is that a correct assumption?

salaboy reviewed (dismissed):
Let's have a discussion to plan for this.. I don't want to embed the Job Executor functionality inside the RuntimeBundle itself.. if we create a starter for it at least we can define if we want to embed it or run it as a separate entity. The end goal, at least in Kubernetes, is to rely on the infrastructure.

igdianov commented:
@Salaboy Let's discuss the use cases to better understand the scope. 

This message based job manager just helps to distribute scheduled executions (async continuations and timers)  among instances of the same runtime bundle via message broker queue worker load balancing. 

According to this [benchmark](http://www.jorambarrez.be/blog/2016/07/20/message-queue-async-executor-benchmark/), MessageBasedJobManager provides 4x better job throughput compared to DefaultJobManager.

daisuke-yoshimoto commented:
@igdianov 
The Activiti Engine has a setting named isMessageQueueMode, and it seems that the default asynchronous job implementation divides behaviors by that setting. Does this fix consider the setting of isMessageQueueMode?

https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/ResetExpiredJobsCmd.java#L36

https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/DefaultAsyncJobExecutor.java#L107

https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/DefaultAsyncJobExecutor.java#L179

daisuke-yoshimoto commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/JobMessageHandler.java:
```diff
@@ -0,0 +1,71 @@
+            
+            // Let's execute job  
+            if(job != null) {
+                ExecuteAsyncRunnable executeAsyncRunnable = new ExecuteAsyncRunnable(job, processEngineConfiguration);
```
@igdianov @salaboy 
Is it okay to generate threads like this directly?

igdianov replied:
@daisuke-yoshimoto It is not creating new thread. ExecuteAsyncRunnable implements Runnable interface that can be executed in another thread or can be just run() in the current thread context.

daisuke-yoshimoto replied:
@igdianov 
I mean that it is common to get new thread from JavaEE's API not directly creating Runnable.

igdianov replied:
@daisuke-yoshimoto Ok. I just wanted to reuse existing logic to execute the job entity resolved from a message. It executes the job synchronously in the message consumer worker thread to be able to throw the exception if the job fails. The message will be redelivered and retried again..

igdianov commented:
> @igdianov
> The Activiti Engine has a setting named isMessageQueueMode, and it seems that the default asynchronous job implementation divides behaviors by that setting. Does this fix consider the setting of isMessageQueueMode?
> 
> https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/ResetExpiredJobsCmd.java#L36
> 
> https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/DefaultAsyncJobExecutor.java#L107
> 
> https://github.com/Activiti/Activiti/blob/develop/activiti-engine/src/main/java/org/activiti/engine/impl/asyncexecutor/DefaultAsyncJobExecutor.java#L179

@daisuke-yoshimoto Yes, it does. See: https://github.com/Activiti/activiti-cloud-runtime-bundle-service/blob/c3927ed81380ed46e1728b4316c2f4ea4f591999/activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java#L78

igdianov committed:
fix: update parent version to 7.1.0-SNAPSHOT

igdianov committed:
feat(JobExecutorIT): add intermediate catching timer event test


fix(JobExecutorIT): use correct intermediateTimerEventExample definition key


fix(JobExecutorIT): add JobExecutorITProcessEngineConfigurer 


fix(MessageBasedJobManager): add @ConditionalOnMissingBean configuration


feat: add MessageBasedJobManagerFactory with default configuration


fix(JobExecutorIT): increase timeout 


fix(JobExecutorIT): remove @Test timeout


fix: try increase timeout


fix(JobExecutorIT): refactor test to reduce timeouts


fix(JobExecutorIT): use TimerJobQuery for assert


fix(JobExecutorIT): add logging


fix(JobExecutorIT): try different settings


feat(JobExecutor): add conditional property for auto configuration


enable test web env

igdianov committed:
chore: polish MessageBasedJobManagerConfigurator

igdianov committed:
feat(JobExecutorIT): add testStartTimeEvent unit test

igdianov committed:
feat(JobExecutorIT): add testBoundaryTimerEvent unit test

igdianov committed:
polish JobExecutorIT assert descriptions

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/JobMessageHandler.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public void handleMessage(Message<?> message) throws MessagingException {
```
I guess that we cannot use StreamListeners here because we need to use dynamic subscription channels, right? Just seeing if we can avoid this cast to get the job id, but I don't think so.

igdianov replied:
Yes, that is the case. We have to bind the channel dynamically and then subscribe/unsubscribe to it in sync with job executor start/stop

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/JobMessageHandler.java:
```diff
@@ -0,0 +1,71 @@
+
+            logger.info("Received job message with id: " + jobId);
+            
+            // Let's try to find existing job by jobId
```
Suggestion: this comment can be replaced by extracting the block below inside a method with a similar name.

igdianov replied:
The message handler execution is a single unit of work that fits into a single method nicely. It will only make sense to externalize `findJobById` and `executeJob` if we need to allow to override JobMessageHandler implementations...

erdemedeiros replied:
Well, when you create a method, it doesn't mean you should reuse it in several places or you need to override it. It can be used only to encapsulate some logic to simplify the caller method (and avoid inline comments). Inline comments are often considered as code smells (unless they are providing information that is not straightforward to get from the code, like this one https://github.com/Activiti/activiti-cloud-runtime-bundle-service/pull/188/files#diff-767a504c8251113456bd793dc386a1a9R52). This blog post covers that subject nicely, please have a look: https://adamtom.at/articles/comments-are-code-smells/.

igdianov replied:
Fixed.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/JobMessageHandler.java:
```diff
@@ -0,0 +1,71 @@
+                }
+              });
+            
+            // Let's execute job  
```
Same here, the comment can be replaced by a method with a similar name.

igdianov replied:
Fixed

igdianov committed:
feat: externalize messageJobConsumerProperties configuration

igdianov committed:
chore: polish JobMessageHandler

igdianov committed:
fix(.gitignore): add .factorypath

igdianov committed:
feat: enable job executor auto-config if missing conditional property

igdianov committed:
fix: add missing @DirtiesContext

igdianov committed:
fix: disable job executor for generic application IT tests

igdianov committed:
fix: TaskIT.adminShouldAssignUser fails intermittently

igdianov committed:
feat: add JobMessageHandlerFactory

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManager.java:
```diff
@@ -0,0 +1,104 @@
+
+    public void sendMessage(final Job jobEntity) {
+        Context.getTransactionContext()
+               .addTransactionListener(TransactionState.COMMITTED, new JobMessageProducer(jobEntity));
```
We need to double check this. If I remember well, the transaction listener was not working as I expected with Spring and I needed to use spring transaction listener to get the right behavior: https://github.com/Activiti/activiti-cloud-runtime-bundle-service/blob/develop/activiti-cloud-services-runtime-bundle/activiti-cloud-services-connectors/src/main/java/org/activiti/services/connectors/IntegrationRequestSender.java#L52.

igdianov replied:
I can refactor `JobMessageProducer` to use `@TransactionalEventListener` to have consistent transaction synchronization approach

igdianov replied:
Fixed.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerAutoConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+public class MessageBasedJobManagerAutoConfiguration {
+    
+    @Bean
+    @ConditionalOnMissingBean
```
Is this `@ConditionalOnMissingBean` safe to be used here? Couldn't we have other beans with the same type for other consumers? Maybe it's necessary to specify the bean name in this case.

igdianov replied:
It is safe. Spring uses bean name derived from method in addition to class to check for condition

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java:
```diff
@@ -0,0 +1,151 @@
+
+public class MessageBasedJobManagerConfigurator implements ProcessEngineConfigurator, SmartLifecycle {
+    
+    public static final String APPLICATION_JSON = "application/json";
```
Let's reuse `MediaType.APPLICATION_JSON_VALUE` from Spring.

igdianov replied:
Yes. I will externalize content type setting as well

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java:
```diff
@@ -0,0 +1,151 @@
+    public void start() {
+        jobMessageHandler = createJobMessageHandler(configuration);
+        
+        // Let's subscribe and bind consumer channel   
```
Can be replaced by a method.

igdianov replied:
It is a part of start/stop unit of work. I think it will only make sense to externalize it if we allow to override configuration behavior. My opinion we should not allow to override it without need.

erdemedeiros replied:
I agree that we should not allow override, but the method is can be private (so not overridable). I'm just concerned about inline comments: if you few the need of adding a comment to describe what a couple of lines is doing, probably you need a (private) method for that.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java:
```diff
@@ -0,0 +1,151 @@
+        inputChannel.subscribe(jobMessageHandler);
+        bindingService.bindConsumer(inputChannel, messageBasedJobManager.getInputChannelName());
+
+        // Now start async executor
```
Can be replaced by a method.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java:
```diff
@@ -0,0 +1,151 @@
+    @Override
+    public void stop() {
+        try {
+            // Let's unbind consumer from input channel
```
Can be replaced by a method.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/MessageBasedJobManagerConfigurator.java:
```diff
@@ -0,0 +1,151 @@
+            bindingService.unbindConsumers(messageBasedJobManager.getInputChannelName());
+            inputChannel.unsubscribe(jobMessageHandler);
+
+            // Let's gracefully shutdown executor
```
Can be replaced by a method.

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    @Test
+    public void testMessageJobConsumerProperties() {
```
Does it make sense to rename the test to use the same pattern used to describe the failure: `shouldConfigureConsumerProperties`?

igdianov replied:
Sure.

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,387 @@
+                                     .isInstanceOf(JobMessageHandler.class);
+    }
+    
+    @SuppressWarnings("deprecation")
```
Shaw we consider to remove `@Deprecated` from `Job` and only mark it as internal? We are using it, after all, so not really deprecated at this point.

igdianov replied:
It makes sense to remove it. I can make a PR in Activiti core

igdianov replied:
See: https://github.com/Activiti/Activiti/pull/2826

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,387 @@
+    public void testAsyncJobs() throws InterruptedException {
+        int jobCount = 100;
+        CountDownLatch jobsCompleted = new CountDownLatch(jobCount);
+        //MessageHandler jobMessageHandlerSpy = Mockito.spy(jobMessageHandler);
```
Please, remove commented code.

igdianov replied:
sure

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,387 @@
+        processEngineConfiguration.getClock().setCurrentTime(new Date(startTime.getTime() + ((5 * 60 * 1000) + 5000)));
+
+        // timer event has been scheduled
+        assertThat(timerScheduled.await(1, TimeUnit.MINUTES)).as("should schedule timer")
```
Do we really need to wait that long?

igdianov replied:
It takes time to schedule timer job especially in slow Travis. It will proceed as soon as latch is triggered maybe within 10 sec, so it is not a big deal to have 1 min timeout.

erdemedeiros replied:
Actually, I think the main reason for the test being failing in Travis is that the configuration defined in the setUp method (https://github.com/Activiti/activiti-cloud-runtime-bundle-service/pull/188/files#diff-134ea6fb38fe95c3f6fa06d9ef418f38R109) is not being taken into account. I fixed the issue in Activiti core and it should be fixed here once this PR is merged
https://github.com/Activiti/activiti-cloud-runtime-bundle-service/pull/326.

I checked locally, all the tests around timers are waiting between 8+ and 9+ seconds. The default timeout for awaitility is 10 seconds, so if Travis is slightly slower that's enough to make the test fail.

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,387 @@
+        processEngineConfiguration.getClock().setCurrentTime(new Date(startTime.getTime() + ((5 * 60 * 1000) + 5000)));
+
+        // timer event has been scheduled
+        assertThat(timerScheduled.await(1, TimeUnit.MINUTES)).as("should schedule timer")
```
Sounds a really long wait.

igdianov replied:
It should proceed as soon as latch is triggered, so it is not really waiting for 1 minute

erdemedeiros reviewed (changes requested):
Great work @igdianov. I've added some comments and questions below.

igdianov committed:
feat: externalize content type configuration

igdianov committed:
fix: use shouldConfigureConsumerProperties method name

igdianov committed:
fix: remove commented code

igdianov committed:
fix: remove `@SuppressWarnings("deprecation")`

Depends on https://github.com/Activiti/Activiti/pull/2826

igdianov committed:
refactor: use @TransactionEventListener in JobMessageProducer

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -0,0 +1,402 @@
+        
+        @Override
+        public void configure(SpringProcessEngineConfiguration processEngineConfiguration) {
+            processEngineConfiguration.setAsyncExecutorDefaultTimerJobAcquireWaitTime(1000);
```
Here, I think we can reduce even more the wait time to impact as less as possible the execution time for this test: `500` ms should be ok.

igdianov replied:
Sounds good. I will try.

igdianov replied:
Fixed

igdianov committed:
fix: polish JobMessageHandler methods

igdianov committed:
fix: reduce AsyncExecutor wait time for tests

igdianov committed:
feat(JobExecutorIT): add testAsyncJobsFailRetry test

igdianov committed:
fix: Codaxy warning

igdianov committed:
feat(JobExecutorIT): add failedTimerJobRetry test

igdianov committed:
fix: add logging to MessageBasedJobManagerConfigurator

igdianov committed:
fix(JobExecutorIT): correct indentation

igdianov committed:
feat: add multiple Rb replicas job executor IT test module

igdianov commented:
@erdemedeiros I had created a separate IT module with multiple Rb replicas to assure job message distribution and execution against shared H2 database instance. The separate IT test module is needed to isolate multiple replicas IT test from Rb Starter test classes classpath conflicts.

igdianov committed:
fix: use conditional on spring.activiti.asyncExecutorActivate property

igdianov committed:
feat: ad MessageBasedJobManager input channel setter

igdianov committed:
feat: add shouldRegisterMessageBasedJobManagerBean test

igdianov committed:
feat: add debug logging to MessageBasedJobManager

igdianov committed:
fix: reduce fail job tests retry interval to 1 second

igdianov committed:
feat(JobMessageProducer): add JobMessageTransactionSynchronization handler

igdianov committed:
fix: correct Codaxy warnings

igdianov commented:
@erdemedeiros I had refactored JobMessageProducer to use Spring TransactionSynchronization handler to synchronize sending of the job messages with resolved message channel. This should minimize the chances of the infrastructure failures. I also implemented publishing application events on job message success and failure with tests.

Question: do we need to add message headers with Rb metadata for job messages?

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java:
```diff
@@ -43,31 +42,57 @@ public DefaultJobMessageProducer(BinderAwareChannelResolver resolver,
+    
+    protected Message<String> buildMessage(Job job) {
+        return MessageBuilder.withPayload(job.getId())
+                // TODO set headers?  
```
Igor, I don't think headers are needed here as the messages are targeting the runtime bundle itself.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java:
```diff
@@ -43,31 +42,57 @@ public DefaultJobMessageProducer(BinderAwareChannelResolver resolver,
-    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
-    public void on(@NonNull JobMessageEvent event) {
-        logger.debug("On JobMessageEvent: {}", event);
+    class JobMessageTransactionSynchronization implements TransactionSynchronization {
```
Is there any specific reason to use `TransactionSynchronization` instead of `@TransactionalEventListener`? I thought `TransactionSynchronization` was the old way of doing it and `@TransactionalEventListener` the new way. In [this blog post](http://pillopl.github.io/reliable-domain-events/), for instance, they are mentioning that code could be simplified by using `@TransactionalEventListener`.

igdianov replied:
I favor a more direct approach to transaction synchronization to be able to fail fast inside main Activiti transaction by validating job attributes and resolving destination channel and then register `TransactionSynchronization` handler to send message after successful commit. The `TransactionalEventListener` uses `TransactionSynchronization` to do the same.

erdemedeiros commented on activiti-cloud-starter-runtime-bundle/src/test/java/org/activiti/cloud/starter/tests/runtime/JobExecutorIT.java:
```diff
@@ -466,6 +518,75 @@ public void testBoundaryTimerEvent() throws Exception {
+                                                             .isTrue();
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
```
Please, prefer `catchThrowable` from AssertJ to check the exception. There is an example here https://github.com/Activiti/activiti-core-common/blob/develop/activiti-spring-connector/src/test/java/org/activiti/core/common/spring/connector/ConnectorDefinitionServiceTest.java#L58-L65

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java:
```diff
@@ -43,31 +42,57 @@ public DefaultJobMessageProducer(BinderAwareChannelResolver resolver,
+                if(!sent)
+                    throw new RuntimeException("Job message cannot be sent due to non-fatal reason from message channel.");
+
+                eventPublisher.publishEvent(new JobMessageSentEvent(job.getId(), destination, job));
```
Who is listening to this event?

igdianov replied:
This is pluggable extension to allow custom application logic to handle job message application events, i.e. logging, error channel DLQ, etc.

igdianov committed:
fix: use assertJ catchThrowable for JobExecutorIT tests

igdianov committed:
fix: correct process instance definition key conflict

igdianov committed:
fix: add missing @DirtiesContext to ITs

igdianov committed:
fix(pom.xml): make spring-cloud-starter-stream-rabbit optional

igdianov committed:
feat: add JobMessage headers

igdianov committed:
chore: add debug logging to JobMessageHandler

igdianov commented:
@erdemedeiros I have added message headers to job messages. This PR is ready to be merged.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/pom.xml:
```diff
@@ -35,6 +35,7 @@
     <dependency>
       <groupId>org.springframework.cloud</groupId>
       <artifactId>spring-cloud-starter-stream-rabbit</artifactId>
+      <optional>true</optional>
```
I'm not a big fun of optional dependencies and I don't think this one is actually needed. I'm going to remove it and see if any problem appears.

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java:
```diff
@@ -25,74 +25,74 @@
     @Override
     public void sendMessage(@NonNull String destination, @NonNull Job job) {
-        Assert.isTrue(TransactionSynchronizationManager.isSynchronizationActive(), "requires active transaction synchronization");
+        if(!TransactionSynchronizationManager.isSynchronizationActive())
```
Please, use brackets for if statements (even if they have a single line).

erdemedeiros commented on activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/JobMessageBuilderAppender.java:
```diff
@@ -0,0 +1,52 @@
+                      .setHeader(JobMessageHeaders.JOB_EXCEPTION_MESSAGE, job.getExceptionMessage())
+                      .setHeader(JobMessageHeaders.JOB_HANDLER_TYPE, job.getJobHandlerType())
+                      .setHeader(JobMessageHeaders.JOB_HANDLER_CONFIGURATION, job.getJobHandlerConfiguration())
+                      .setHeader(JobMessageHeaders.JOB_TENANT_ID, job.getTenantId())
```
We are not exposing tenants, for now, let's remove it from the headers.

erdemedeiros committed:
Remove dependency to rabbit

erdemedeiros committed:
Add brackets to if statements

erdemedeiros committed:
Remove tenantId header

erdemedeiros committed:
Fix Codacy warnings

erdemedeiros commented:
Fixes https://github.com/Activiti/Activiti/issues/2863
